clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -469,73 +426,190 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
+                }
+            } else {
+                prevAssignment.put(partition, new ConsumerGenerationPair(consumer, generation));
+            }
+        }
+    }
+
+    /**
+     * filling in the prevAssignment from the subscriptions.
+     *
+     * @param subscriptions:        Map from the member id to their respective topic subscription
+     * @param prevAssignment:       The assignment contains the assignment with the 2nd largest generation
+     */
     private void prepopulateCurrentAssignments(Map<String, Subscription> subscriptions,
-                                               Map<String, List<TopicPartition>> currentAssignment,
                                                Map<TopicPartition, ConsumerGenerationPair> prevAssignment) {
         // we need to process subscriptions' user data with each consumer's reported generation in mind
         // higher generations overwrite lower generations in case of a conflict
         // note that a conflict could exists only if user data is for different generations
 
-        // for each partition we create a sorted map of its consumers by generation
-        Map<TopicPartition, TreeMap<Integer, String>> sortedPartitionConsumersByGeneration = new HashMap<>();
         for (Map.Entry<String, Subscription> subscriptionEntry: subscriptions.entrySet()) {
             String consumer = subscriptionEntry.getKey();
-            MemberData memberData = memberData(subscriptionEntry.getValue());
-
-            for (TopicPartition partition: memberData.partitions) {
-                if (sortedPartitionConsumersByGeneration.containsKey(partition)) {
-                    Map<Integer, String> consumers = sortedPartitionConsumersByGeneration.get(partition);
-                    if (memberData.generation.isPresent() && consumers.containsKey(memberData.generation.get())) {
-                        // same partition is assigned to two consumers during the same rebalance.
-                        // log a warning and skip this record
-                        log.warn("Partition '{}' is assigned to multiple consumers following sticky assignment generation {}.",
-                            partition, memberData.generation);
-                    } else
-                        consumers.put(memberData.generation.orElse(DEFAULT_GENERATION), consumer);
-                } else {
-                    TreeMap<Integer, String> sortedConsumers = new TreeMap<>();
-                    sortedConsumers.put(memberData.generation.orElse(DEFAULT_GENERATION), consumer);
-                    sortedPartitionConsumersByGeneration.put(partition, sortedConsumers);
-                }
+            Subscription subscription = subscriptionEntry.getValue();
+            if (subscription.userData() != null) {
+                // since this is our 2nd time to deserialize memberData, rewind userData is necessary
+                subscription.userData().rewind();
             }
```

Comment:
This is actually a bug after `constrainedAssign` implemented. After `constrainedAssign` implemented, we'll do `allSubscriptionsEqual` to decide if we want to use `constrainedAssign` or `generalAssign`. In `allSubscriptionsEqual`, we not only check if subscription equal, but also deserialize the user data. So, if it is deserialized once, the position of userData (ByteBuffer) will be moved to the end of the buffer, so that we have to rewind here.